### ISSUE TITLE
ERNIE3.0分类模型，需要270多个类的分类模型，直接训的话，在保存参数的时候直接超出内存的问题

### ISSUE THREAD

**Issue by jiweizhangxu** [PaddlePaddle/PaddleNLP#2623]
我现在需要270多个类的分类模型，直接训的话，在保存参数的时候直接超出内存，内存大概32G，咨询了微信群，重新试了试还是一样，挂上就占了23G，参数保存时提高到了23.95G如图：
![1](https://user-images.githubusercontent.com/88191580/175244842-78f9278b-db76-459f-b946-c09319753b76.png)
![3](https://user-images.githubusercontent.com/88191580/175244857-5371d50d-d051-4c81-ae9f-1cbd8d2fdfc7.png)

有什么好办法吗？

**Comment by jiweizhangxu** [comment]
代码如下：
import os
import paddle
import paddlenlp
# 自定义数据集
import re
from paddlenlp.transformers import BertForSequenceClassification, BertTokenizer
from paddlenlp.transformers import ErnieForSequenceClassification, ErnieTokenizer
from paddlenlp.datasets import load_dataset
from paddlenlp.transformers import export_model
import argparse

def parse_args():
    parser = argparse.ArgumentParser()

    # Required parameters
    parser.add_argument(
        "--model_type",
        default=None,
        type=str,
        required=True,
        help="Model type selected in the list: " +
        ", ".join(MODEL_CLASSES.keys()), )
    parser.add_argument(
        "--model_path",
        default=None,
        type=str,
        required=True,
        help="Path of the trained model to be exported.", )
    parser.add_argument(
        "--output_path",
        default=None,
        type=str,
        required=True,
        help="The output file prefix used to save the exported inference model.",
    )
    args = parser.parse_args()
    return args

def clean_text(text):
    text = text.replace("\r", "").replace("\n", "")
    text = re.sub(r"\\n\n", ".", text)
    return text

# 定义读取数据集函数
def read_custom_data(is_test=False, is_one_hot=True):

    file_num = 6 if is_test else 48
    # filepath = 'raw_data/test/' if is_test else 'raw_data/train/'
    # with open('data_2/divorce')
    filepath = 'data_300/for_1234_300.txt' if is_test else 'data_300/text.txt'

    for i in range(file_num):
        f = open(filepath)
        while True:
            line = f.readline()
            if not line:
                break
            data = line.strip().split('\t')
            # 标签用One-hot表示
            if is_one_hot:
                labels = [float(1) if str(i) in data[1].split(',') else float(0) for i in range(281)]
            else:
                labels = [int(d) for d in data[1].split(',')]
            yield {"text": clean_text(data[0]), "labels": labels}
        f.close()
#这里具体类别有点不方便提供，有281个类
label_vocab = {}

# load_dataset()创建数据集
train_ds = load_dataset(read_custom_data, is_test=False, lazy=False)
test_ds = load_dataset(read_custom_data, is_test=True, lazy=False)

# lazy=False，数据集返回为MapDataset类型
print("数据类型:", type(train_ds))

# labels为One-hot标签
print("训练集样例:", train_ds[0])
print("测试集样例:", test_ds[0])

from paddlenlp.transformers import AutoModelForSequenceClassification, AutoTokenizer

model_name = "ernie-3.0-base-zh"
num_classes = 281
model = AutoModelForSequenceClassification.from_pretrained(model_name, num_classes=num_classes)
tokenizer = AutoTokenizer.from_pretrained(model_name)


import functools
import numpy as np

from paddle.io import DataLoader, BatchSampler
from paddlenlp.data import DataCollatorWithPadding

# 数据预处理函数，利用分词器将文本转化为整数序列
def preprocess_function(examples, tokenizer, max_seq_length):
    result = tokenizer(text=examples["text"], max_seq_len=max_seq_length)
    result["labels"] = examples["labels"]
    return result

trans_func = functools.partial(preprocess_function, tokenizer=tokenizer, max_seq_length=128)
train_ds = train_ds.map(trans_func)
test_ds = test_ds.map(trans_func)

# collate_fn函数构造，将不同长度序列充到批中数据的最大长度，再将数据堆叠
collate_fn = DataCollatorWithPadding(tokenizer)

# 定义BatchSampler，选择批大小和是否随机乱序，进行DataLoader
train_batch_sampler = BatchSampler(train_ds, batch_size=16, shuffle=True)
test_batch_sampler = BatchSampler(test_ds, batch_size=16, shuffle=False)
train_data_loader = DataLoader(dataset=train_ds, batch_sampler=train_batch_sampler, collate_fn=collate_fn)
test_data_loader = DataLoader(dataset=test_ds, batch_sampler=test_batch_sampler, collate_fn=collate_fn)

#数据训练和评估
import time
import paddle.nn.functional as F

from metric import MultiLabelReport

# Adam优化器、交叉熵损失函数、自定义MultiLabelReport评价指标
optimizer = paddle.optimizer.AdamW(learning_rate=1e-4, parameters=model.parameters())
criterion = paddle.nn.BCEWithLogitsLoss()
metric = MultiLabelReport()

prog = paddle.static.default_main_program()

from eval import evaluate

epochs = 20  # 训练轮次
ckpt_dir = "ernie_ckpt_1234_300/model_all"  # 训练过程中保存模型参数的文件夹
CUDA_VISIBLE_DEVICES = 1
global_step = 0  # 迭代次数
tic_train = time.time()
best_f1_score = 0

MODEL_CLASSES = {
    "bert": (BertForSequenceClassification, BertTokenizer),
    "ernie": (ErnieForSequenceClassification, ErnieTokenizer),
}

for epoch in range(1, epochs + 1):
    for step, batch in enumerate(train_data_loader, start=1):
        input_ids, token_type_ids, labels = batch['input_ids'], batch['token_type_ids'], batch['labels']

        # 计算模型输出、损失函数值、分类概率值、准确率、f1分数
        logits = model(input_ids, token_type_ids)
        loss = criterion(logits, labels)
        probs = F.sigmoid(logits)
        metric.update(probs, labels)
        auc, f1_score, _, _ = metric.accumulate()

        # 每迭代10次，打印损失函数值、准确率、f1分数、计算速度
        global_step += 1
        if global_step % 10 == 0:
            print(
                "global step %d, epoch: %d, batch: %d, loss: %.5f, auc: %.5f, f1 score: %.5f, speed: %.2f step/s"
                % (global_step, epoch, step, loss, auc, f1_score,
                   10 / (time.time() - tic_train)))
            tic_train = time.time()

        # 反向梯度回传，更新参数
        loss.backward()
        optimizer.step()
        optimizer.clear_grad()

        #每迭代40次，评估当前训练的模型、保存当前最佳模型参数和分词器的词表等
        if global_step % 20 == 0:
            save_dir = ckpt_dir
            if not os.path.exists(save_dir):
                os.makedirs(save_dir)
            eval_f1_score = evaluate(model, criterion, metric, test_data_loader, label_vocab, if_return_results=False)
            if eval_f1_score > best_f1_score:
                best_f1_score = eval_f1_score
                model.save_pretrained(save_dir)
                tokenizer.save_pretrained(save_dir)
                model_class, tokenizer_class = MODEL_CLASSES["ernie"]



##########################################################################################################
#多标签分类预测结果预测
from eval import evaluate

# 模型在测试集中表现
model.set_dict(paddle.load('ernie_ckpt_1234_300/model_all/model_state.pdparams'))

# 也可以选择加载预先训练好的模型参数结果查看模型训练结果
# model.set_dict(paddle.load('ernie_ckpt_trained/model_state.pdparams'))

print("ERNIE 3.0 在12345多标签分类test集表现", end= " ")
results = evaluate(model, criterion, metric, test_data_loader, label_vocab)

test_ds = load_dataset(read_custom_data, is_test=True, is_one_hot=False, lazy=False)
res_dir = "./results"
if not os.path.exists(res_dir):
    os.makedirs(res_dir)
with open(os.path.join(res_dir, "multi_label.tsv"), 'w', encoding="utf8") as f:
    f.write("text\tprediction\n")
    for i, pred in enumerate(results):
        f.write(test_ds[i]['text']+"\t"+pred+"\n")

**Comment by LiuChiachi** [comment]
感谢关注:)
23G 不是比 32G小了吗
如果需要进一步减小内存占用，可以使用梯度累计方法用小batch_size模拟大batch_size，可以参考这里的代码
https://github.com/PaddlePaddle/PaddleNLP/blob/0315365dbafa6e3b1c7147121ba85e05884125a5/model_zoo/ernie-3.0/run_seq_cls.py#L287-L288

https://github.com/PaddlePaddle/PaddleNLP/blob/0315365dbafa6e3b1c7147121ba85e05884125a5/model_zoo/ernie-3.0/run_seq_cls.py#L386-L393
